File: Quick Sort/quickSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>  

using namespace std;

int partition(vector<int>& arr, int low, int high) {
    int index = low - 1;
    int p = low + rand() % (high - low + 1);
    int pivot = arr[p];
    swap(arr[high], arr[p]);

    for (int i = low; i <= high; i++) {
        if (arr[i] <= pivot) {
            swap(arr[++index], arr[i]);
        }
    }

    return index;
}

void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int p = partition(arr, low, high);

        quickSort(arr, low, p - 1);
        quickSort(arr, p + 1, high);
    }
}

int main() {
    vector<int> arr = {3, 1, 4, 1, 15, 9, 2, 6, 5, 3, 5,20};
    int size = arr.size();

    quickSort(arr, 0, size - 1);

    cout << "Sorted array: ";
    for (const auto& element : arr) {
        cout << element << " ";
    }
    cout << endl;

    return 0;
}

File: Quick Sort/quickSort.py
# import random

# def partition(arr, low, high):
#     pivot_index = random.randint(low,high-1)
#     pivot = arr[pivot_index]

#     arr[pivot_index], arr[high] = arr[high], arr[pivot_index]

#     index = low - 1
#     for i in range(low, high):
#         if arr[i] < pivot:
#             index += 1
#             arr[i], arr[index] = arr[index], arr[i]

#     arr[index + 1], arr[high] = arr[high], arr[index + 1]
#     return index + 1

# def quickSort(arr, low, high):
#     if high > low:
#         p = partition(arr, low, high)
#         quickSort(arr, low, p - 1)
#         quickSort(arr, p + 1, high)


# arr = [12, 21, 3, 12, 3, 1, 2]
# quickSort(arr, 0, len(arr) - 1)
# print(arr)


import random

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    else:
        # Choose a random pivot
        pivot_index = random.randint(0, len(arr) - 1)
        pivot = arr[pivot_index]

        # Partition the array into elements less than and greater than the pivot
        less = [x for i, x in enumerate(arr) if x <= pivot and i != pivot_index]
        greater = [x for i, x in enumerate(arr) if x > pivot and i != pivot_index]

        # Recursively sort the sub-arrays
        return quicksort(less) + [pivot] + quicksort(greater)

# Example usage:
my_list = [12, 21, 3, 12, 3, 1, 2]
sorted_list = quicksort(my_list)
print(sorted_list)